File: internal/bazi/derived.go
package bazi

import "github.com/6tail/tyme4go/tyme"

import "slices"

// hideStemTypes tyme 的藏干层次枚举到本模块类型的映射，下标即枚举值
var hideStemTypes = [...]HideStemType{
	tyme.RESIDUAL: HideStemResidual,
	tyme.MIDDLE:   HideStemMiddle,
	tyme.MAIN:     HideStemMain,
}

// TenStarOf 某个天干对日主的十神
func TenStarOf(dayStem tyme.HeavenStem, target tyme.HeavenStem) string {
	return dayStem.GetTenStar(target).GetName()
}

// HideStemsOf 地支的藏干，按本气、中气、余气排列
func HideStemsOf(dayStem tyme.HeavenStem, branch tyme.EarthBranch) []HideStem {
	hides := branch.GetHideHeavenStems()
	out := make([]HideStem, 0, len(hides))
	for _, h := range hides {
		stem := h.GetHeavenStem()
		out = append(out, HideStem{
			Stem:    stem.GetName(),
			Type:    hideStemTypes[h.GetType()],
			TenStar: TenStarOf(dayStem, stem),
			Element: stem.GetElement().GetName(),
		})
	}
	return out
}

// BranchTenStarOf 地支本气藏干对日主的十神，大运流年格里的地支十神取这一位
func BranchTenStarOf(dayStem tyme.HeavenStem, branch tyme.EarthBranch) string {
	return TenStarOf(dayStem, branch.GetHideHeavenStemMain())
}

// BuildPillar 组装一柱
//
// emptyBranches 是日柱旬空的两个地支，用于标注本柱地支是否落空
func BuildPillar(
	kind PillarKind,
	cycle tyme.SixtyCycle,
	dayStem tyme.HeavenStem,
	emptyBranches []string,
) Pillar {
	stem := cycle.GetHeavenStem()
	branch := cycle.GetEarthBranch()

	extras := cycle.GetExtraEarthBranches()
	extraNames := make([]string, 0, len(extras))
	for _, b := range extras {
		extraNames = append(extraNames, b.GetName())
	}

	stemTenStar := "日主"
	if kind != PillarDay {
		stemTenStar = TenStarOf(dayStem, stem)
	}

	return Pillar{
		Kind:          kind,
		SixtyCycle:    cycle.GetName(),
		Stem:          stem.GetName(),
		Branch:        branch.GetName(),
		StemElement:   stem.GetElement().GetName(),
		BranchElement: branch.GetElement().GetName(),
		StemTenStar:   stemTenStar,
		HideStems:     HideStemsOf(dayStem, branch),
		Sound:         cycle.GetSound().GetName(),
		Terrain:       dayStem.GetTerrain(branch).GetName(),
		SelfTerrain:   stem.GetTerrain(branch).GetName(),
		Ten:           cycle.GetTen().GetName(),
		ExtraBranches: extraNames,
		Empty:         slices.Contains(emptyBranches, branch.GetName()),
	}
}

// PillarOf 按柱取值
func PillarOf(pillars Pillars, kind PillarKind) Pillar {
	switch kind {
	case PillarYear:
		return pillars.Year
	case PillarMonth:
		return pillars.Month
	case PillarDay:
		return pillars.Day
	case PillarHour:
		return pillars.Hour
	}
	return Pillar{}
}

File: web/core/src/bazi/derived.ts
/**
 * 柱的派生项：十神、藏干、纳音、十二长生、旬空
 *
 * 这五张表 `tyme4ts` 都已内置，本文件只做组装与命名映射
 */

import {
  EarthBranch,
  HeavenStem,
  HideHeavenStemType,
  SixtyCycle,
} from "tyme4ts"

import type {
  FiveElement,
  HideStem,
  HideStemType,
  Pillar,
  PillarKind,
} from "./types"

const HIDE_TYPE_MAP: Record<number, HideStemType> = {
  [HideHeavenStemType.MAIN]: "main",
  [HideHeavenStemType.MIDDLE]: "middle",
  [HideHeavenStemType.RESIDUAL]: "residual",
}

/** 某个天干对日主的十神，日主自身记作 `日主` */
export function tenStarOf(dayStem: HeavenStem, target: HeavenStem): string {
  return dayStem.getTenStar(target).getName()
}

/** 地支的藏干，按本气、中气、余气排列 */
export function hideStemsOf(
  dayStem: HeavenStem,
  branch: EarthBranch
): HideStem[] {
  return branch.getHideHeavenStems().map((h) => {
    const stem = h.getHeavenStem()
    return {
      stem: stem.getName(),
      type: HIDE_TYPE_MAP[h.getType()],
      tenStar: tenStarOf(dayStem, stem),
      element: stem.getElement().getName() as FiveElement,
    }
  })
}

/** 地支本气藏干对日主的十神，大运流年格里的地支十神取这一位 */
export function branchTenStarOf(
  dayStem: HeavenStem,
  branch: EarthBranch
): string {
  return tenStarOf(dayStem, branch.getHideHeavenStemMain())
}

/**
 * 组装一柱
 *
 * @param emptyBranches 日柱旬空的两个地支，用于标注本柱地支是否落空
 */
export function buildPillar(
  kind: PillarKind,
  cycle: SixtyCycle,
  dayStem: HeavenStem,
  emptyBranches: readonly string[]
): Pillar {
  const stem = cycle.getHeavenStem()
  const branch = cycle.getEarthBranch()
  return {
    kind,
    sixtyCycle: cycle.getName(),
    stem: stem.getName(),
    branch: branch.getName(),
    stemElement: stem.getElement().getName() as FiveElement,
    branchElement: branch.getElement().getName() as FiveElement,
    stemTenStar: kind === "day" ? "日主" : tenStarOf(dayStem, stem),
    hideStems: hideStemsOf(dayStem, branch),
    sound: cycle.getSound().getName(),
    terrain: dayStem.getTerrain(branch).getName(),
    selfTerrain: stem.getTerrain(branch).getName(),
    ten: cycle.getTen().getName(),
    extraBranches: cycle.getExtraEarthBranches().map((b) => b.getName()),
    empty: emptyBranches.includes(branch.getName()),
  }
}
